### PR TITLE
feat: toHaveAccessibleName matcher

### PR DESCRIPTION
**Pull request by anishamalde**
Currently, this test fails as I am unsure on how to extract the text from the nested Text Element

```
test('getByLabelText supports nested aria-labelledby', async () => {
  const screen = render(
    <>
      <View nativeID="label">
        <Text>Nested Aria LabelledBy</Text>
      </View>
      <TextInput testID="text-input" aria-labelledby="label" />
    </>
  );

  const element = screen.getByTestId('text-input');
  expect(element).toHaveAccessibleName('Nested Aria LabelledBy');
});
```

### Summary

https://github.com/callstack/react-native-testing-library/issues/1503 - Implement toHaveAccessibileName matcher

### Test plan
Add unit tests and ensure all test passes

### CODE REVIEW

File: src/helpers/accessiblity.ts [hunk 1]
```diff
@@ -233,3 +233,34 @@ export function isElementSelected(
...
+        return label;
+      }
+    }
+  }
+
+  if (labelTextFromLabel) {
+    return labelTextFromLabel;
+  }
+
+  const ownText = element?.children?.join(' ');
```
Comment (mdjastrzebski) [inline src/helpers/accessiblity.ts:263]: There is a `getTextContent` fn already avialable in RNTL that would do that for you.

File: src/helpers/accessiblity.ts [hunk 2]
```diff
@@ -233,3 +233,34 @@ export function isElementSelected(
...
 }
+
+export function getAccessibleName(
+  element: ReactTestInstance
+): string | undefined {
+  const labelTextFromLabel = getAccessibilityLabel(element);
+  const labelTextFromLabelledBy = getAccessibilityLabelledBy(element);
+
+  if (labelTextFromLabelledBy) {
+    const parentElement = getHostParent(element);
```
Comment (mdjastrzebski) [inline src/helpers/accessiblity.ts:244]: labelled by does not have to be necessarily nested under the same parent. You should look from root element. simllar to: https://github.com/callstack/react-native-testing-library/blob/a9f32a69e46bedc9413dd0bdf74637f7a3b2b579/src/helpers/matchers/matchLabelText.ts#L40

You can access current root with `screen.UNSAFE_root`

File: src/matchers/to-have-accessible-name.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  element: ReactTestInstance,
+  expectedName?: TextMatch,
+  options?: TextMatchOptions
+) {
+  checkHostElement(element, toHaveAccessibleName, this);
+
+  const receivedName = getAccessibleName(element);
+
+  return {
+    pass: matchAccessibleName(
```
Comment (mdjastrzebski) [inline src/matchers/to-have-accessible-name.tsx:19]: [nit] move match accessible name to inside this file.

File: src/matchers/extend-expect.ts [hunk 4]
```diff
@@ -33,4 +34,5 @@ expect.extend({
   toHaveProp,
   toHaveStyle,
   toHaveTextContent,
+  toHaveAccessibleName,
```
Comment (mdjastrzebski) [inline src/matchers/extend-expect.ts:37]: [nit] sort alphabetically

File: src/matchers/extend-expect.ts [hunk 5]
```diff
@@ -15,6 +15,7 @@ import { toHaveDisplayValue } from './to-have-display-value';
 import { toHaveProp } from './to-have-prop';
 import { toHaveStyle } from './to-have-style';
 import { toHaveTextContent } from './to-have-text-content';
+import { toHaveAccessibleName } from './to-have-accessible-name';
```
Comment (mdjastrzebski) [inline src/matchers/extend-expect.ts:18]: [nit] sort alphabetically

File: src/matchers/extend-expect.d.ts [hunk 6]
```diff
@@ -20,6 +20,7 @@ export interface JestNativeMatchers<R> {
   toHaveProp(name: string, expectedValue?: unknown): R;
   toHaveStyle(style: StyleProp<Style>): R;
   toHaveTextContent(expectedText: TextMatch, options?: TextMatchOptions): R;
+  toHaveAccessibleName(expectedName?: TextMatch, options?: TextMatchOptions): R;
```
Comment (mdjastrzebski) [inline src/matchers/extend-expect.d.ts:23]: [nit] sort alphabetically

File: src/matchers/__tests__/to-have-accessible-name.test.tsx [hunk 7]
```diff
@@ -0,0 +1,111 @@
+import * as React from 'react';
+import { View, Text, TextInput } from 'react-native';
+import { render, screen } from '../..';
+import '../extend-expect';
+
+test('toHaveAccessibleName() on view with accessibilityLabel prop', () => {
+  render(<View testID="accessibility-label" accessibilityLabel="Test Label" />);
+  const element = screen.getByTestId('accessibility-label');
+  expect(element).toHaveAccessibleName('Test Label');
```
Comment (mdjastrzebski) [inline src/matchers/__tests__/to-have-accessible-name.test.tsx:9]: Also check if `.toHavaAccessibleName()` (without name)-  it should verify if the name is present regardles of actual value.

Add also negative cases, e.g. `not.toHaveAccessibleName("Other Label")

**Comment by mdjastrzebski** [comment]
rebased on main

**Comment by mdjastrzebski** [comment]
@pierrezimmermannbam yeah, that's my plan, but been pretty busy recently, I plan to get this done this week in order to have release full Jest matchers feature (this is the last matcher! 🥳). BTW would you have a time to take a look at proposed docs in #1506 PR for Jest matchers docs?

**Review by mdjastrzebski — APPROVED** [review]
All good! Time to merge!

**Comment by pierrezimmermannbam** [comment]
Nice ! I'll take a look at the docs PR today

**Comment by mdjastrzebski** [comment]
🎉 Release in version: [v12.4.0](https://github.com/callstack/react-native-testing-library/releases/tag/v12.4.0)

**Pull request merged**